main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala

```diff
@@ -0,0 +1,1081 @@
+        val resTpe = tvar
+        val resEff = eff
+        (resTpe, resEff)
+
+      case Expr.ArrayLit(exps, exp, tvar, evar, loc) =>
+        val regionVar = Type.freshVar(Kind.Eff, loc)
+        val regionType = Type.mkRegion(regionVar, loc)
+        val (tpes, effs) = exps.map(visitExp).unzip
+        val (tpe, eff) = visitExp(exp)
+        c.expectTypeM(expected = regionType, actual = tpe, exp.loc)
+        val elmTpe = c.unifyAllTypesM(tpes, Kind.Star, loc)
+        c.unifyTypeM(tvar, Type.mkArray(elmTpe, regionVar, loc), loc)
+        c.unifyTypeM(evar, Type.mkUnion(Type.mkUnion(effs, loc), eff, regionVar, loc), loc)
+        val resTpe = tvar
+        val resEff = evar
+        (resTpe, resEff)
+
+      case Expr.ArrayNew(exp1, exp2, exp3, tvar, evar, loc) =>
+        val regionVar = Type.freshVar(Kind.Eff, loc)
+        val regionType = Type.mkRegion(regionVar, loc)
+        val (tpe1, eff1) = visitExp(exp1)
+        val (tpe2, eff2) = visitExp(exp2)
+        val (tpe3, eff3) = visitExp(exp3)
+        c.expectTypeM(expected = regionType, actual = tpe1, loc)
+        c.expectTypeM(expected = Type.Int32, actual = tpe3, exp3.loc)
+        c.unifyTypeM(tvar, Type.mkArray(tpe2, regionVar, loc), loc)
+        c.unifyTypeM(evar, Type.mkUnion(eff1, eff2, eff3, regionVar, loc), loc)
+        val resTpe = tvar
+        val resEff = evar
+        (resTpe, resEff)
+
+      case Expr.ArrayLoad(exp1, exp2, tvar, evar, loc) =>
+        val regionVar = Type.freshVar(Kind.Eff, loc)
+        val (tpe1, eff1) = visitExp(exp1)
+        val (tpe2, eff2) = visitExp(exp2)
+        c.expectTypeM(expected = Type.mkArray(tvar, regionVar, loc), actual = tpe1, exp1.loc)
+        c.expectTypeM(expected = Type.Int32, actual = tpe2, exp2.loc)
+        c.unifyTypeM(evar, Type.mkUnion(regionVar, eff1, eff2, loc), loc)
+        val resTpe = tvar
+        val resEff = evar
+        (resTpe, resEff)
+
+      case Expr.ArrayStore(exp1, exp2, exp3, evar, loc) =>
+        val elmVar = Type.freshVar(Kind.Star, loc)
+        val regionVar = Type.freshVar(Kind.Eff, loc)
+        val arrayType = Type.mkArray(elmVar, regionVar, loc)
+        val (tpe1, eff1) = visitExp(exp1)
+        val (tpe2, eff2) = visitExp(exp2)
+        val (tpe3, eff3) = visitExp(exp3)
+        c.expectTypeM(expected = arrayType, actual = tpe1, exp1.loc)
+        c.expectTypeM(expected = Type.Int32, actual = tpe2, exp2.loc)
+        c.expectTypeM(expected = elmVar, actual = tpe3, exp3.loc)
+        c.unifyTypeM(evar, Type.mkUnion(List(regionVar, eff1, eff2, eff3), loc), loc)
+        val resTpe = Type.Unit
+        val resEff = evar
+        (resTpe, resEff)
+
+      case Expr.ArrayLength(exp, loc) =>
+        val elmVar = Type.freshVar(Kind.Star, loc)
+        val regionVar = Type.freshVar(Kind.Eff, loc)
+        val (tpe, eff) = visitExp(exp)
+        c.expectTypeM(Type.mkArray(elmVar, regionVar, loc), tpe, exp.loc)
+        val resTpe = Type.Int32
+        val resEff = eff
```

Comment:
I almost fell into the trap of suggesting a union was missing, maybe a comment is warranted?

Reply:
I think if I add a comment then Magnus will tell me to remove it :P